### PR TITLE
[IOTDB-2740] Equal size bucket sampling UDFs: EQUAL_SIZE_BUCKET_RANDOM_SAMPLE, EQUAL_SIZE_BUCKET_AGG_SAMPLE, EQUAL_SIZE_BUCKET_M4_SAMPLE

### PR DESCRIPTION
**Pull request by ZhanGHanG9991**
## Description
Add EqualBucketSample in UDF

<hr>

This PR has:
- [x] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: docs/zh/UserGuide/Query-Data/Select-Expression.md [hunk 1]
```diff
@@ -529,6 +529,142 @@ select s1, zero_count(s1), non_zero_count(s2), zero_duration(s3), non_zero_durat
...
+```
+
+#### 等分桶聚合采样
+
+当`method`参数为`aggregation`时，采用聚合采样法对输入序列进行采样。当选择`aggregation`作为采样方法，用户需要另外提供一个聚合函数参数即
+- `type`：聚合类型，取值为`avg`或`max`或`min`。在缺省情况下，采用`avg`。
+
+| 函数名      | 可接收的输入序列类型                     | 必要的属性参数                               | 输出序列类型     | 功能类型                                             |
+|----------|--------------------------------|---------------------------------------|------------|--------------------------------------------------|
+| EQUAL_BUCKET_SAMPLE   | INT32 / INT64 / FLOAT / DOUBLE | `method`:`aggregation`</br>`proportion`取值范围为`(0, 1]`，默认为`0.1`</br>`type`:取值类型有`avg`, `max`, `min`, 默认为`avg`  | INT32 / INT64 / FLOAT / DOUBLE | 返回符合采样比例的等分桶聚合采样                |
```
Comment (SteveYurongSu) [inline docs/zh/UserGuide/Query-Data/Select-Expression.md:608]: sum, extreme, variance 

Those sampling methods are all acceptable!

File: docs/zh/UserGuide/Query-Data/Select-Expression.md [hunk 2]
```diff
@@ -529,6 +529,142 @@ select s1, zero_count(s1), non_zero_count(s2), zero_duration(s3), non_zero_durat
...
+|1970-01-01T08:00:00.080+08:00|             84.5|   89.0|   80.0|
+|1970-01-01T08:00:00.090+08:00|             94.5|   99.0|   90.0|
++-----------------------------+-----------------+-------+-------+
+Total line number = 10
+It costs 0.037s
+
+```
+#### 等分桶M4采样
+
+当`method`参数为`M4`时，采用M4采样法对输入序列进行采样。即对于每个桶采样首、尾、最小和最大值。
```
Comment (SteveYurongSu) [inline docs/zh/UserGuide/Query-Data/Select-Expression.md:639]: 建议对首、尾、和除去首尾外的最小和最大值进行采样，这样点数不会比预期少

File: docs/zh/UserGuide/Query-Data/Select-Expression.md [hunk 3]
```diff
@@ -529,6 +529,142 @@ select s1, zero_count(s1), non_zero_count(s2), zero_duration(s3), non_zero_durat
 +-----------------------------+-------------+-------------------------+-----------------------------+----------------------------+--------------------------------+
 ```
 
+### 等分桶降采样函数
```
Comment (SteveYurongSu) [inline docs/zh/UserGuide/Query-Data/Select-Expression.md:532]:
```suggestion
### 等数量分桶降采样函数
```

Comment (SteveYurongSu) [reply docs/zh/UserGuide/Query-Data/Select-Expression.md:532]: 把其他类似的名字改成“等数量分桶”吧

File: docs/zh/UserGuide/Query-Data/Select-Expression.md (hunk 1 above)
Comment (ZhanGHanG9991) [reply docs/zh/UserGuide/Query-Data/Select-Expression.md:608]: I've added these methods.

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketSample.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  protected TSDataType dataType;
+  protected String aggMethodType;
+  protected double proportion;
+  protected int bucketSize;
+
+  @Override
+  public void validate(UDFParameterValidator validator) throws MetadataException, UDFException {
+    dataType = validator.getParameters().getDataType(0);
+    proportion = validator.getParameters().getDoubleOrDefault("proportion", 0.1);
+    aggMethodType = validator.getParameters().getStringOrDefault("type", "avg").toLowerCase();
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketSample.java:41]: aggMethodType 的验证放到 server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java 更合适一些

Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketSample.java:41]: aggMethodType 的验证放到 server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java 更合适一些

Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketSample.java]: 好的，已经修改提交

File: docs/zh/UserGuide/Query-Data/Select-Expression.md (hunk 3 above)
Comment (ZhanGHanG9991) [reply docs/zh/UserGuide/Query-Data/Select-Expression.md:532]: 好的，文档已更新

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+    @Override
+    public void aggregateInt(RowWindow rowWindow, PointCollector collector) throws IOException {
+      long time = rowWindow.getRow(0).getTime();
+      int windowSize = rowWindow.windowSize();
+
+      double sum = 0;
+      for (int i = 0; i < windowSize; i++) {
+        sum += rowWindow.getRow(i).getInt(0) * 1.0 / windowSize;
+      }
+      collector.putDouble(time, sum);
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java:62]:
```suggestion
      double sum = 0;
      for (int i = 0; i < windowSize; i++) {
        sum += rowWindow.getRow(i).getInt(0);
      }
      collector.putDouble(time, sum / windowSize);
```

File: docs/zh/UserGuide/Query-Data/Select-Expression.md [hunk 6]
```diff
@@ -529,6 +529,146 @@ select s1, zero_count(s1), non_zero_count(s2), zero_duration(s3), non_zero_durat
...
+|1970-01-01T08:00:00.086+08:00|         86.0|
+|1970-01-01T08:00:00.096+08:00|         96.0|
++-----------------------------+-------------+
+Total line number = 10
+It costs 0.024s
+```
+
+#### 等数量分桶聚合采样
+
+采用聚合采样法对输入序列进行采样，用户需要另外提供一个聚合函数参数即
```
Comment (SteveYurongSu) [inline docs/zh/UserGuide/Query-Data/Select-Expression.md:602]: 说一下每个桶输出的时间的行为（每个桶采样输出的时间戳为第一个点的时间戳）

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+      int windowSize = rowWindow.windowSize();
+
+      double avg = 0, sum = 0;
+      for (int i = 0; i < windowSize; i++) {
+        avg += rowWindow.getRow(i).getInt(0) * 1.0 / windowSize;
+      }
+      for (int i = 0; i < windowSize; i++) {
+        sum += (rowWindow.getRow(i).getInt(0) - avg) * (rowWindow.getRow(i).getInt(0) - avg);
+      }
+      collector.putDouble(time, sum / windowSize);
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java:355]:
```suggestion
      double avg = 0, sum = 0;
      
      for (int i = 0; i < windowSize; i++) {
        avg += rowWindow.getRow(i).getInt(0);
      }
      avg /= windowSize;
      
      for (int i = 0; i < windowSize; i++) {
        int delta = rowWindow.getRow(i).getInt(0) - avg;
        sum += delta * delta;
      }
      
      collector.putDouble(time, sum / windowSize);
```

File: docs/zh/UserGuide/Query-Data/Select-Expression.md (hunk 6 above)
Comment (ZhanGHanG9991) [reply docs/zh/UserGuide/Query-Data/Select-Expression.md:602]: 好的，文档已修改

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java (hunk 5 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java:62]: 已修改该部分实现

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java (hunk 7 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketAggSample.java]: 已修改这部分的实现

**Review by SteveYurongSu — APPROVED** [review]
LGTM

**Pull request merged**